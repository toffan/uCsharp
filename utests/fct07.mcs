typedef struct {int x; int y;} Point;
typedef struct {Point* c; int r;} Cercle;
typedef struct {Point a; Point b;} Segment;

int* malloc(int size) {
    int* addr;
    asm {
        LOAD %size
        SUBR Malloc
        STORE %addr
    }
    return addr;
}

void memset(int* p, int val) {
    *p = val;
}

int main() {
    int* p = malloc(1);
    memset(p, 9);
    Point* pt = (Point*) malloc(2);
    memset((int*) pt, 8);
    memset((int*) (((int) pt) + 1), 11);
    memset(p, pt->y);

    return 0;
}
