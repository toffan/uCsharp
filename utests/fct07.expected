JUMP __malloc
_malloc:
    PUSH 1
	LOAD (1) -1[LB]
	SUBR Malloc
	STORE (1) 3[LB]
    LOAD (1) 3[LB]
    RETURN (1) 1
    RETURN (0) 1
__malloc:

JUMP __memset
_memset:
    LOAD (1) -1[LB]
    LOAD (1) -1[ST]
    LOAD (1) -2[LB]
    STOREI (1)
    POP (0) 1
    RETURN (0) 2
__memset:

JUMP __main
_main:
    LOADL 1
    CALL (SB) _malloc
    LOAD (1) 3[LB]
    LOADL 9
    CALL (SB) _memset
    LOADL 2
    CALL (SB) _malloc
    LOAD (1) 4[LB]
    LOADL 8
    CALL (SB) _memset
    LOAD (1) 4[LB]
    LOADL 1
    SUBR IAdd
    LOADL 11
    CALL (SB) _memset
    LOAD (1) 3[LB]
    LOAD (1) 4[LB]
    LOADL 1
    SUBR IAdd
    LOADI (1)
    CALL (SB) _memset

    LOADL 0
    RETURN (1) 0
    RETURN (0) 0
__main:

    CALL (SB) _main
    HALT
